src/phpDocumentor/Dsn.php

```diff
@@ -0,0 +1,288 @@
+            throw new \InvalidArgumentException(
+                sprintf('"%s" is not a valid DSN.', $dsn)
+            );
+        }
+
+        $this->parseScheme($locationParts);
+
+        $this->parseHostAndPath($locationParts);
+
+        $this->parsePort($locationParts);
+
+        $this->user = isset($locationParts['user']) ? $locationParts['user'] : "";
+
+        $this->password = isset($locationParts['pass']) ? $locationParts['pass'] : "";
+
+        $this->parseQuery($locationParts);
+
+        $this->parseParameters($dsnParts);
+    }
+
+    /**
+     * validates and sets the scheme property
+     *
+     * @param string $locationParts
+     * @return void
+     */
+    private function parseScheme($locationParts)
+    {
+        $validSchemes = ['file', 'git+http', 'git+https'];
+        if (! in_array(strtolower($locationParts['scheme']), $validSchemes)) {
+            throw new \InvalidArgumentException(
+                sprintf('"%s" is not a valid scheme.', $locationParts['scheme'])
+            );
+        }
+        $this->scheme = strtolower($locationParts['scheme']);
+    }
+
+    /**
+     * Validates and sets the host and path properties
+     *
+     * @param string $locationParts
+     * @return void
+     */
+    private function parseHostAndPath($locationParts)
+    {
+        $path = isset($locationParts['path']) ? $locationParts['path'] : "";
+        $host = isset($locationParts['host']) ? $locationParts['host'] : "";
+
+        if ($this->getScheme() === 'file') {
+            $this->path = $host . $path;
+        } else {
+            $this->host = $host;
+            $this->path = $path;
+        }
+    }
+
+    /**
+     * Validates and sets the port property
+     *
+     * @param string $locationParts
+     * @return void
+     */
+    private function parsePort($locationParts)
+    {
+        if (! isset($locationParts['port'])) {
+            if ($this->getScheme() === 'git+http') {
+                $this->port = 80;
+            } elseif ($this->getScheme() === 'git+https') {
+                $this->port = 443;
+            } else {
+                $this->port = 0;
+            }
+        } else {
+            $this->port = $locationParts['port'];
+        }
+    }
+
+    /**
+     * validates and sets the query property
+     *
+     * @param string $locationParts
+     * @return void
+     */
+    private function parseQuery($locationParts)
+    {
+        if (isset($locationParts['query'])) {
+            $queryParts = explode('&', $locationParts['query']);
+            foreach ($queryParts as $part) {
+                $option = $this->splitKeyValuePair($part);
+
+                $this->query[$option[0]] = $option[1];
+            }
+        }
+    }
+
+    /**
+     * validates and sets the parameters property
+     *
+     * @param string $dsnParts
+     * @return void
+     */
+    private function parseParameters($dsnParts)
+    {
+        if (!empty($dsnParts)) {
```

Comment:
This check is not necessary if $dsnParts is an array; if $dsnParts is empty then the foreach will never be entered